Update gtfs_headway_analysis.py

carlhiggs opened the pull request:
Updated to account for frequencies.txt, and use folder of assets with process/data/GTFS (ie. unified data folder with rest of project)   --- that latter change may not be desireable, but the other one is important.  Also includes fixes with taking mean across dates, rather than minimum; and creating a 'direction_id' field, and ensuring this does not have nulls, to avoid if else conditions related to this.

carlhiggs committed:
Update gtfs_headway_analysis.py

Updated to account for frequencies.txt, and use folder of assets with process/data/GTFS (ie. unified data folder with rest of project)   --- that latter change may not be desireable, but the other one is important.  Also includes fixes with taking mean across dates, rather than minimum; and creating a 'direction_id' field, and ensuring this does not have nulls, to avoid if else conditions related to this.

carlhiggs committed:
Update gtfs_headway_analysis.py

Addressed issues in set_date_service_table(loaded_feeds) with integration of calendar_dates (optional) within calendar (optional) --- the main observed issue (causing problems with reading Barcelona) was that weekdays were not given to additional dates added into date_service_df using calendar_dates.  The function was also amended to allow for possibility that calendar.txt may not exist, and if so loaded_feeds.calendar would be a null data frame with no columns.  This would cause the function to fail, even if dates were defined in calendar_dates.  With the amended function, 2364 bus stops and 150 train stops are loaded for the 'gtfs_spain_barcelona_TMB_20190704' feed.

carlhiggs committed:
Update gtfs_headway_analysis.py

Fixed typo in previous amendment (trip_routes >>> trips_routes)

carlhiggs committed:
Update gtfs_config.py

updated GTFS config with new feeds for Barcelona and Valencia (will upload to cloudstor)

shiqin-liu commented on process/GTFS_freq_stop/gtfs_headway_analysis.py:
```diff
@@ -311,151 +342,173 @@ def get_hlc_stop_frequency(loaded_feeds, start_hour, end_hour, start_date,
         loaded_feeds = ua_load.gtfsfeed_to_df(gtfsfeed_path=gtfsfeed_path, validation=True, bbox=bbox, remove_stops_outsidebbox=True)
-
+        
+        for root, dirs, files in os.walk(gtfsfeed_path):
```
So does that mean if the frequency file exists, that would be our preferred option to use for calculating headway? I think it makes sense, as mentioned in tidytransit, it is more accurate than using stop_time alone? The code looks good, just want to make sure that's what we are aiming for :).

shiqin-liu commented on process/GTFS_freq_stop/gtfs_headway_analysis.py:
```diff
@@ -311,151 +342,173 @@ def get_hlc_stop_frequency(loaded_feeds, start_hour, end_hour, start_date,
-
+        
         # load GTFS Feed
         loaded_feeds = ua_load.gtfsfeed_to_df(gtfsfeed_path=gtfsfeed_path, validation=True, bbox=bbox, remove_stops_outsidebbox=True)
```
We need to either update the bbox information in config with new boudaries, or take out the bbox option `ua_load.gtfsfeed_to_df(gtfsfeed_path=gtfsfeed_path)` to load all stops.

shiqin-liu commented:
Copying your notes from email here:

> - bring in frequencies.txt (urban access doesn't do this it seems)
> - simplified some of the datetime parsing (allowed some things to be written more concisely)
> - added direction_id, and ensured there were no nulls (to allow grouping over this, avoid logic loop - less code)
> - ensure headways are averaged across dates, and then the lowest average headway in either direction is taken 
> - comment out the 'representative week' code -- this code caused Mexico City trains (i think, something like that) not to be processed.  We average over dates anyway, so I think this is fair. 
> - weight hours in frequency.txt  (for taking weighted average of headway for hours within analysis window; hours outside analysis window are weighted zero --- but before then, any zero weight rows are dropped anyway)
> - if there are stops with valid trip frequencies and positive weight defined, these will be used; any other stops are defined using stop_times if possible

Thanks Carl! I think these changes all look good. Sorry that I am not able to test out these on my only laptop as I am running our process code in master branch. But if it works at yours, it should be good! I will wait for your confirmation to merge this PR :).

carlhiggs commented:
Thanks Shirley --- I'll aim to look at this more later today (maybe we could specify input boundary location in config to automatically derive bbox?); there's an issue with input data for a couple of other cities though, so I'll look at that first.  Hope that's okay, Carl.

carlhiggs committed:
Update gtfs_config.py

Updated bounding box coordinates for 500 metres buffer to cities' current urban study region boundaries (EPSG:4326 WGS 84)

carlhiggs committed:
Update gtfs_headway_analysis.py

Updated to output layers for each city with feeds combined; each layer has a new 'feed' attribute to describe the feed from which the stop was sourced.  This means the layers are more directly usable in the global indicators project (subject to filtering by headway).  Also replaced some formatting with f strings to simplify print statements.
